### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by dmitchelldm74**
added create account on index [ / ]
next: add password

### CODE REVIEW

File: db_access.py [hunk 1]
```diff
@@ -1,9 +1,18 @@
 import sqlite3
-
+import uuid
 from flask import app, g
 
 from chirpr import app
 
+def add_chirp(text):
```
Comment (ewilson) [inline db_access.py]: It looks like this is not used at this time, please remove it for now.

**Comment by ewilson** [comment]
The largest missing piece here is the password. When a new user signs up they should choose a password. This means that you will have to add a column to the database (create a migration script, `ALTER TABLE` etc. Please ask if you aren't sure about this.

Also, the trello card mentioned explicitly that an error message should display if the username has been taken. This did not work for me. Here's a link about flash error messages in Flask: http://flask.pocoo.org/docs/0.11/patterns/flashing/

I see that you have added `jquery.js` to the repository. Please remove that. We can either use a CDN, or we can use bower for managing dependencies, but external dependencies should not be in the repo.

**Comment by ewilson** [comment]
Test comment

File: db_access.py [hunk 2]
```diff
@@ -1,9 +1,23 @@
 import sqlite3
-
+import uuid, datetime
```
Comment (ewilson) [inline db_access.py]: remove uuid

File: db_access.py [hunk 3]
```diff
@@ -1,9 +1,23 @@
...
+    num = -1
+    for c in conn.execute('SELECT id FROM chirp'):
+        num = c[0]
+    num += 1
+    conn.execute('''
+        INSERT INTO chirp VALUES (?,?,?,?)
+    ''', (num, text, 1, str(datetime.datetime.utcnow())))
+    conn.commit()
+    return num
+    #c.id, c.body, c.datetime, u.handle FROM chirp c, user u WHERE c.user_id = u.id
```
Comment (ewilson) [inline db_access.py]: This method is overly complicated, and inefficient. You wouldn't want to iterate over every `id` every time you add a chirp.

Look at `add_user`, and note that it doesn't specify the id -- which is totally OK, the DB handles it. This is the effect of `AUTOINCREMENT` in the `CREATE TABLE` statement.

File: templates/account.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+{% block body %}
+<h1>Create an Account:</h1>
+<form action="/admin/user/add" method="POST">
+    <label>Username: </label><input type="text" name="handle" class="form-control"/><br/>
+    <label>Password: </label><input type="password" name="password" class="form-control"/><br/>
+    <input type="submit" class="btn btn-block btn-warning" value="Submit"><br/>
+</form>
+<script>
+function id(elid){return document.getElementById(elid);}
+</script>
```
Comment (ewilson) [inline templates/account.html]: I don't understand why you are adding this. I see that it is a convenience method for `getElementById` but I don't see it used.

File: templates/chirp.html [hunk 5]
```diff
@@ -0,0 +1,17 @@
+{% extends "base.html" %}
```
Comment (ewilson) [inline templates/chirp.html]: We need to be disciplined about the scope of the card. When working on a card about logins, it will be important not to introduce unrelated functionality, such as chirping.

File: chirpr.py [hunk 6]
```diff
@@ -20,13 +20,42 @@ def users():
...
 def add_user():
     handle = request.form.get('handle')
-    db_access.add_user(handle)
+    password = request.form.get('password')
+    data = db_access.add_user(handle, password)
+    if data != None:
+        return redirect(url_for('account', success='0'))
+    tup = db_access.sign_in(handle, password)
+    if tup[0] == True:
+        session['user'] = tup[1]
```
Comment (ewilson) [inline chirpr.py]: The card we are working on is signing up. The sign-in card is next, so should not be in this pull request. Please remove sign-in functionality.

File: chirpr.py [hunk 7]
```diff
@@ -20,13 +20,42 @@ def users():
...
     return redirect(url_for('users'))
+  
+
+@app.route('/account')
+def account():
+    return render_template('account.html')
+ 
+ 
+@app.route('/user/login', methods=['POST'])
+def login():
```
Comment (ewilson) [inline chirpr.py]: `login` should not be part of this card, please remove it.

File: chirpr.py [hunk 8]
```diff
@@ -35,11 +64,37 @@ def chirps():
...
-
-
+    if 'user' in session:
+        db_access.delete_chirp(chirp_id, session['user'])
+        return redirect(url_for('chirp'))
+    return redirect(url_for('index'))
+    
+
+@app.errorhandler(404)
+def page_not_found(error):
```
Comment (ewilson) [inline chirpr.py]: Errorhandling would be another example of scope-creep.

File: db_access.py [hunk 9]
```diff
@@ -1,21 +1,42 @@
 import sqlite3
-
+import uuid, datetime, hashlib
```
Comment (ewilson) [inline db_access.py]: I'm not sure why `uuid` is back.

**Comment by ewilson** [comment]
You are implementing things quickly, as usual. However, this pull request will not be approved until it fits with the features outlined on the card.

Having pull requests for small pieces of functionality allows me to help you to build things well. When features are added very quickly, I am not able to help you think about architecture, testability, performance, design patterns, etc.

File: templates/account.html (hunk 4 above)
Comment (ewilson) [reply templates/account.html]: This `id` function should be removed, as it is not used.

Comment (eric-aver) [reply templates/account.html]: OK, I see that it is used, but I would recommend that you not use custom functions that duplicate built-in JS functions.

File: chirpr.py [hunk 10]
```diff
@@ -20,26 +18,36 @@ def users():
...
-@app.route('/admin/chirp/delete/<chirp_id>')
+    
+    
+@app.route('/chirp/delete/<chirp_id>')
 def delete_chirp(chirp_id):
-    db_access.delete_chirp(chirp_id)
-    return redirect(url_for('chirps'))
-
+    if 'user' in session:
+        db_access.delete_chirp(chirp_id, session['user'])
```
Comment (ewilson) [inline chirpr.py]: Please remove these changes, they are not part of creating a new account.

File: db_access.py [hunk 11]
```diff
@@ -1,21 +1,27 @@
...
 from flask import app, g
 
 from chirpr import app
 
-def get_all_chirps():
+def hash_ps(text):
+    t = hashlib.sha224()
+    t.update(text.encode('utf-8'))
+    t = t.hexdigest()
+    return t
```
Comment (ewilson) [inline db_access.py]: My approach would be:

```
def hash_ps(text):
    return hashlib.sha224(password.encode('utf-8')).hexdigest()
```

Calling `update` would be typical if you were iterating through a loop and building the string to be hashed. In this case, it doesn't seem to add much. Also, your code raises the question, "what exactly is `t`?"

The answer is, `t` is a hash instance -- until it is a hashed string. Your code here works fine, but it is not best to use the same variable to refer to very different things.

File: db_access.py [hunk 12]
```diff
@@ -1,21 +1,27 @@
...
     conn = get_db()
-    return conn.execute('''
-        SELECT c.id, c.body, c.datetime, u.handle
-        FROM chirp c, user u
-        WHERE c.user_id = u.id
-    ''').fetchall()
+    chirps = []
+    for text in conn.execute('SELECT c.id, c.body, c.datetime, u.handle FROM chirp c, user u, followers f WHERE f.handleid = ? AND c.user_id = u.id AND c.user_id = f.followerid ORDER BY c.id', (uid,)):
+        chirps.append(text)
+    return reversed(chirps)
```
Comment (ewilson) [inline db_access.py:16]: I don't believe there is a `followers` table yet -- if you have made a DB migration for that, it didn't get checked in.

Also, don't use Python to reverse the order of the results from the DB. Write the query in a way to get the results in the order that you want. (Use `order by foo desc` if you want them backwards.

Finally, this is not related to creating new users, so remove it for now.

File: db_access.py [hunk 13]
```diff
@@ -1,21 +1,27 @@
...
 
 
-def delete_chirp(chirp_id):
+def delete_chirp(chirp_id, user_id):
     conn = get_db()
-    conn.execute('DELETE FROM chirp WHERE id = :id', {'id': chirp_id})
+    conn.execute('DELETE FROM chirp WHERE id = :id AND user_id = :uid', {'id': chirp_id, 'uid':user_id})
```
Comment (ewilson) [inline db_access.py]: This doesn't need to be changed as part of the current card

File: db_access.py [hunk 14]
```diff
@@ -24,16 +30,25 @@ def get_all_users():
     return conn.execute('SELECT id, handle, admin FROM user').fetchall()
 
 
+def user_for(uid):
+    conn = get_db()
+    for a in conn.execute('SELECT handle FROM user WHERE id=:id', {'id':uid}):
+        return a[0]
+    return ''
```
Comment (ewilson) [inline db_access.py]: I don't think this is needed for the current card, please let me know if I'm missing something important.

File: db_access.py [hunk 15]
```diff
@@ -24,16 +30,25 @@ def get_all_users():
...
-def add_user(handle):
+def add_user(handle, password):
+    password = hash_ps(password)
     conn = get_db()
-    conn.execute('INSERT INTO user (handle, admin) values (:handle, :admin)',
-                 {'handle': handle, 'admin': 0})
+    for c in conn.execute('SELECT * FROM user WHERE handle=:handle', {'handle':handle}):
+        return list(c)
+    conn.execute('INSERT INTO user (handle, password, admin) values (:handle, :password, :admin)',
+                 {'handle': handle, 'password':password, 'admin': 0})
```
Comment (ewilson) [inline db_access.py:40]: You could just call the `hash_ps` method in the dictionary:

```
  {'handle': handle, 'password':hash_ps(password), 'admin': 0}
```

But if you would rather call it on it's own line, please name the variable something else, like `hashed_ps`

File: db_access.py [hunk 16]
```diff
@@ -49,6 +64,12 @@ def connect_db():
     return conn
 
 
+def sign_in(handle, password):
+    conn = get_db()
+    password = hash_ps(password)
+    for c in conn.execute('SELECT id FROM user WHERE handle=:handle AND password=:password', {'handle':handle, 'password':password}):
+        return (True, c[0])
+    return (False, -1)
```
Comment (ewilson) [inline db_access.py]: Again, leave out any login logic until creating users is complete.

File: templates/account.html [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    <font color='red' id='msg'></font><br>
+    <label>Username: </label><input type="text" name="handle" class="form-control"/><br/>
+    <label>Password: </label><input type="password" name="password" class="form-control"/><br/>
+    <input type="submit" class="btn btn-block btn-warning" value="Submit"><br/>
+</form>
+<script>
+var state = '{{ request.args.get("success") }}';
+if(state == 'create0'){
+    id('msg').innerHTML = 'You cannot create an account with this username.';
+}
```
Comment (ewilson) [inline templates/account.html]: I wasn't explicit about this, when I said:

> Flash error message should display if username already exists in db

I was thinking of Flask's `flash` functionality. Please read this and consider implementing that way. (If you want to do it using JS, tell me why you think that is better.

**Comment by ewilson** [comment]
In addition to the above comments, I notice that when I click the `Chirp` button on the toolbar, I get a 404 page.

If a page doesn't exist yet, do not display a dead link on the page.

There are one or two more things that need addressed, but it is getting too late. I'll try to address them by tomorrow evening. But this should give you enough to work on in the meantime.

File: chirpr.py [hunk 18]
```diff
@@ -20,26 +18,36 @@ def users():
...
-@app.route('/admin/user/add', methods=['POST'])
+@app.route('/user/add', methods=['POST'])
 def add_user():
     handle = request.form.get('handle')
-    db_access.add_user(handle)
+    password = request.form.get('password')
+    data = db_access.add_user(handle, password)
+    if data != None:
+        return redirect(url_for('account', success='create0'))
     return redirect(url_for('users'))
```
Comment (ewilson) [inline chirpr.py:39]: It is bad practice to make successful DB actions indicated by a return value of `None`. We need to rethink this.

File: db_access.py (hunk 15 above)
Comment (ewilson) [reply db_access.py:40]: Also, returning a list of users from an `add_user` method is not at all expected, we need a better way.

My suggestion: create a method `get_user_by_handle` that tries to retrieve a user. Then call `add_user` (from `chirpr.py` if no users are found, flash an error otherwise.

File: templates/account.html (hunk 17 above)
Comment (ewilson) [reply templates/account.html]: I'm thinking about this again, and I think you should use Flask's `flash`: http://flask.pocoo.org/docs/0.11/patterns/flashing/

Main advantage is that you take advantage of an existing, tested solution for the problem of errors, instead of creating your own.

File: .gitignore [hunk 19]
```diff
@@ -58,7 +58,7 @@ coverage.xml
 
 # Django stuff:
 *.log
-
+*.sh
```
Comment (ewilson) [inline .gitignore:61]: I'm curious why you put this in the `.gitignore`. If you created a shell script that helps you, you don't want to lose it.

The repo should contain all files you want for development -- even if the deployed application doesn't need them.

File: chirpr.py [hunk 20]
```diff
@@ -20,26 +18,41 @@ def users():
...
+    data = False
+    error = None
+    if (len(handle) < 20 and len(handle) > 1) == (False):
+        error = 'Username is too %s.' % ('long' if len(handle) > 1 else 'short')
+    elif (len(password) < 20 and len(password) > 4) == (False):
+        error = 'Password is too %s.' % ('long' if len(password) > 1 else 'short')
+    if error is None:
+        data = db_access.create_account(handle, password)
+    if data is False:
+        flash(error, 'account_err')
```
Comment (ewilson) [inline chirpr.py:37]: The error message `'account_err'` could be friendlier. Something like:

> Username already exists, not created

File: db_access.py [hunk 21]
```diff
@@ -1,19 +1,22 @@
...
+
+def get_all_chirps(uid):
     conn = get_db()
-    return conn.execute('''
-        SELECT c.id, c.body, c.datetime, u.handle
-        FROM chirp c, user u
-        WHERE c.user_id = u.id
-    ''').fetchall()
+    chirps = []
+    for text in conn.execute('SELECT c.id, c.body, c.datetime, u.handle FROM chirp c, user u, followers f WHERE c.user_id = u.id ORDER BY c.id'):
```
Comment (ewilson) [inline db_access.py]: You still have a reference to `followers` here, that needs to be removed at this time.

File: db_access.py [hunk 22]
```diff
@@ -30,13 +33,25 @@ def delete_user(user_id):
...
+    conn.execute('INSERT INTO user (handle, password, admin) values (:handle, :password, :admin)',
+                 {'handle': handle, 'password':password, 'admin': 0})
     conn.commit()
 
 
+def create_account(handle, password):
+    conn = get_db()
+    res = conn.execute('SELECT * FROM user WHERE handle=:handle', {'handle':handle})
+    if list(res):
+        state = False
```
Comment (ewilson) [inline db_access.py:48]: You _could_ `flash` from here, if you want. Not required, just a thought.

File: db_access.py [hunk 23]
```diff
@@ -49,6 +64,11 @@ def connect_db():
     return conn
 
 
+def sign_in(handle, password):
```
Comment (ewilson) [inline db_access.py]: Sign in, not yet! We are very close.

File: templates/account.html [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+{% block body %}
+<h1>Create an Account:</h1>
+<form action="/user/add" method="POST">
+    {% with errors = get_flashed_messages(category_filter=["account_err"]) %}
+        {% if errors %}
+            {%- for msg in errors %}
+                <font color='red' id='msg'>{{ msg }}</font>
+            {% endfor -%}
+        {% endif %}  
+    {% endwith %}
```
Comment (ewilson) [inline templates/account.html]: This message flashing should be in `base.html` -- that way it will be available on _all_ pages.

**Comment by ewilson** [comment]
Here's what happens when I try to run it:

``` ~/code/chirpr
ericwilson [20:17:56] $ python3 chirpr.py
Traceback (most recent call last):
  File "chirpr.py", line 3, in <module>
    import db_access
  File "/Users/ericwilson/code/chirpr/db_access.py", line 5, in <module>
    from chirpr import app
  File "/Users/ericwilson/code/chirpr/chirpr.py", line 56, in <module>
    app.secret_key = open('SECRET_KEY~', 'r').read().replace('\n', '')
FileNotFoundError: [Errno 2] No such file or directory: 'SECRET_KEY~'
```

Since you don't need the `SECRET_KEY` stuff for this PR, you could remove all references to it. Otherwise, you could use an environment variable, not a file.

File: db_access.py (hunk 22 above)
Comment (dmitchelldm74) [reply db_access.py:48]: I think keeping the flash in the chirpr.py file is the best way to go. Everything flash in one file.
